#471 - Add ILM documentation (#474)
jaegertracing/documentation: content/docs/next-release/deployment.md

To learn more about rollover index management in Jaeger refer to this
[article](https://medium.com/jaegertracing/using-elasticsearch-rollover-to-manage-indices-8b3d0c77915d).

> #### Elasticsearch ILM support

Author: I had similar thought initially - but dint act on it. Agreed, will move it.

{{< warning >}}
Experimental Feature, Added in release 1.22.

Supported versions: 7.x
{{< /warning >}}
[Elasticsearch ILM](https://www.elastic.co/guide/en/elasticsearch/reference/current/index-lifecycle-management.html) helps us automatically manage indices according to our performance, resiliency, and retention requirements.
For example, you could use ILM to:

* Spin up a new index when an index reaches a certain size or number of documents
* Create a new index each day, week, or month and archive previous ones
* Delete stale indices to enforce data retention standards

> 

Reviewer: I would aim to make the steps below as simple as possible for users to follow. A suggestion is to provide code that they can copy-paste where practical, even if it's just an example that they can modify later on, and avoid indirection. Having all the steps in one place just for ILM means we can avoid touching any existing manual rollover documentation.

I've provided a suggestion below.

ILM usage for jaeger indices can be enabled by following below steps:

- Create an ILM policy in elasticsearch named *jaeger-ilm-policy*
- Run elasticsearch initializer with ES_USE_ILM=true in environment variables. (details in below section)

> - After the initialization, deploy Jaeger with `--es.use-ilm=true` and `--es.use-aliases=true`.

Reviewer: I propose we replace this block with (and remove the changes from L336 onwards):

##### Enabling ILM support

1. Create an ILM policy in elasticsearch named *jaeger-ilm-policy*.

   For example, the following policy will rollover the "active" index when it is
   older than 1m and delete indices that are older than 2m.
   ```
   curl -X PUT http://localhost:9200/_ilm/policy/jaeger-ilm-policy \
   -H 'Content-Type: application/json; charset=utf-8' \
   --data-binary @- << EOF
   {
     "policy": {
       "phases": {
         "hot": {
           "min_age": "0ms",
           "actions": {
             "rollover": {
               "max_age": "1m"
             },
             "set_priority": {
               "priority": 100
             }
           }
         },
         "delete": {
           "min_age": "2m",
           "actions": {
             "delete": {}
           }
         }
       }
     }
   }
   EOF
   ```
2. Run elasticsearch initializer with `ES_USE_ILM=true`:
   ```
   docker run -it --rm --net=host -e ES_USE_ILM=true jaegertracing/jaeger-es-rollover:latest init http://localhost:9200 # <1>
   ```
   <1> If you need to initialize archive storage, add `-e ARCHIVE=true`.

   {{< info >}}
   While initializing with ILM support, make sure that an ILM policy named *jaeger-ilm-policy* is created in Elasticsearch beforehand.
   Initialization checks for this policy - if it does not exist, it would error out with following message.

   *ILM policy jaeger-ilm-policy doesn't exist in Elasticsearch. Please create it and rerun init*
   {{< /info >}}
3. After the initialization, deploy Jaeger with `--es.use-ilm=true` and `--es.use-aliases=true`.

Author: Agreed, adding it. Just wanted mention something with respect to above ILM policy to be accurate, even though the max age in above policy is set to 1m,  rollover would only happen after 10m - as indices.lifecycle.poll_interval is set to 10 mins by default. 
This can be changed in cluster settings though. :)

<1> Remove indices older than 14 days.

#### Elasticsearch ILM support
{{< warning >}}
Experimental feature added in [release v1.22.0](https://github.com/jaegertracing/jaeger/releases/tag/v1.22.0).

Supported Elasticsearch versions: 7.x
{{< /warning >}}
[Elasticsearch ILM](https://www.elastic.co/guide/en/elasticsearch/reference/current/index-lifecycle-management.html) automatically manages indices according to performance, resiliency, and retention requirements.

For example:
* Rollover to a new index by size (bytes or number of documents) or age, archiving previous indices
* Delete stale indices to enforce data retention standards

###### Enabling ILM support
* Create an ILM policy in elasticsearch named jaeger-ilm-policy.

>     For example, the following policy will rollover the "active" index when it is
>     older than 1m and delete indices that are older than 2m.

Reviewer: ```suggestion
  For example, the following policy will rollover the "active" index when it is
  older than 1m and delete indices that are older than 2m.
```

  ```shell

>     curl -X PUT http://localhost:9200/_ilm/policy/jaeger-ilm-policy \
>     -H 'Content-Type: application/json; charset=utf-8' \
>     --data-binary @- << EOF
>     {

Reviewer: ```suggestion
  curl -X PUT http://localhost:9200/_ilm/policy/jaeger-ilm-policy \
  -H 'Content-Type: application/json; charset=utf-8' \
  --data-binary @- << EOF
  {
```

    "policy": {
      "phases": {
        "hot": {
          "min_age": "0ms",
          "actions": {
            "rollover": {
              "max_age": "1m"
            },
            "set_priority": {
              "priority": 100
            }
          }
        },
        "delete": {
          "min_age": "2m",
          "actions": {
            "delete": {}
          }
        }
      }
    }
  }
    EOF

>     ```

Reviewer: unindent: line up with above `}`, otherwise the `EOF` will not work.

Author: Yep, my bad. Done and checked the curl myself.

* Run elasticsearch initializer with `ES_USE_ILM=true`:

  ```shell
    docker run -it --rm --net=host -e ES_USE_ILM=true jaegertracing/jaeger-es-rollover:latest init http://localhost:9200 # <1>

>     ```

Reviewer: unindent by two spaces in line with above opening code block.

  <1> If you need to initialize archive storage, add `-e ARCHIVE=true`.

  {{< info >}}
  While initializing with ILM support, make sure that an ILM policy named `jaeger-ilm-policy` is created in Elasticsearch beforehand (see the previous step), otherwise the following error message will be shown:

  "ILM policy jaeger-ilm-policy doesn't exist in Elasticsearch. Please create it and rerun init"
  {{< /info >}}

>    After the initialization, deploy Jaeger with `--es.use-ilm=true` and `--es.use-aliases=true`.

Reviewer: ```suggestion
  After the initialization, deploy Jaeger with `--es.use-ilm=true` and `--es.use-aliases=true`.
```

{{< /warning >}}
[Elasticsearch ILM](https://www.elastic.co/guide/en/elasticsearch/reference/current/index-lifecycle-management.html) automatically manages indices according to performance, resiliency, and retention requirements.

For example:
* Rollover to a new index by size (bytes or number of documents) or age, archiving previous indices
* Delete stale indices to enforce data retention standards

###### Enabling ILM support
* Create an ILM policy in elasticsearch named jaeger-ilm-policy.

  For example, the following policy will rollover the "active" index when it is
  older than 1m and delete indices that are older than 2m.

```shell
  curl -X PUT http://localhost:9200/_ilm/policy/jaeger-ilm-policy \
  -H 'Content-Type: application/json; charset=utf-8' \
  --data-binary @- << EOF
  {
  "policy": {
    "phases": {
      "hot": {
        "min_age": "0ms",
        "actions": {
          "rollover": {
            "max_age": "1m"
          },
          "set_priority": {
            "priority": 100
          }
        }
      },
      "delete": {
        "min_age": "2m",
        "actions": {
          "delete": {}
        }
      }
    }
  }
}
  EOF
  ```

> * Run elasticsearch initializer with ES_USE_ILM=true:

Reviewer: ```suggestion
* Run elasticsearch initializer with `ES_USE_ILM=true`:
```

> ```shell

Reviewer: ```suggestion
  ```shell
```

  docker run -it --rm --net=host -e ES_USE_ILM=true jaegertracing/jaeger-es-rollover:latest init http://localhost:9200 # <1>
  ```
<1> If you need to initialize archive storage, add -e ARCHIVE=true.

> {{< info >}}
> While initializing with ILM support, make sure that an ILM policy named jaeger-ilm-policy is created in Elasticsearch beforehand.
> Initialization checks for this policy - if it does not exist, it would error out with following message.
> 
> "ILM policy jaeger-ilm-policy doesn't exist in Elasticsearch. Please create it and rerun init"
> {{< /info >}}

Reviewer: ```suggestion
  {{< info >}}
  While initializing with ILM support, make sure that an ILM policy named `jaeger-ilm-policy` is created in Elasticsearch beforehand (see the previous step), otherwise the following error message will be shown:

  "ILM policy jaeger-ilm-policy doesn't exist in Elasticsearch. Please create it and rerun init"
  {{< /info >}}
```
